h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/SparseMatrixFactory.java

```diff
@@ -0,0 +1,366 @@
+            SparseMatrix matrix, SparseMatrixDimensions dimensions,
+            Vec.Reader respVec, float[] resp, float[] weights
+        ) {
+            _actualRows = new int[chunks.length];
+            
+            _f = f;
+            _vecs = vecs;
+            _w = w;
+            _di = di;
+            _matrix = matrix;
+            _dims = dimensions;
+            _respVec = respVec;
+            _resp = resp;
+            _weights = weights;
+        }
+
+        @Override
+        protected void map(int chunk) {
+            long nonZeroCount = _dims._precedingNonZeroElementsCounts[chunk];
+            int rwRow = _dims._precedingRowCounts[chunk];
+            NestedArrayPointer rowHeaderPointer = new NestedArrayPointer(_dims._precedingRowCounts[chunk]);
+            NestedArrayPointer dataPointer = new NestedArrayPointer(nonZeroCount);
+
+            for(long i = _f.anyVec().espc()[chunk]; i < _f.anyVec().espc()[chunk+1]; i++) {
+                if (_w != null && _w.at(i) == 0) continue;
+                rowHeaderPointer.setAndIncrement(_matrix._rowHeaders, nonZeroCount);
+                _actualRows[chunk]++;
+                for (int j = 0; j < _di._cats; ++j) {
+                    dataPointer.set(_matrix._sparseData, 1);
+                    if (_vecs[j].isNA(i)) {
+                        dataPointer.set(_matrix._colIndices, _di.getCategoricalId(j, Float.NaN));
+                    } else {
+                        dataPointer.set(_matrix._colIndices, _di.getCategoricalId(j, _vecs[j].at8(i)));
+                    }
+                    dataPointer.increment();
+                    nonZeroCount++;
+                }
+                for (int j = 0; j < _di._nums; ++j) {
+                    float val = (float) _vecs[_di._cats + j].at(i);
+                    if (val != 0) {
+                        dataPointer.set(_matrix._sparseData, val);
+                        dataPointer.set(_matrix._colIndices, _di._catOffsets[_di._catOffsets.length - 1] + j);
+                        dataPointer.increment();
+                        nonZeroCount++;
+                    }
+                }
+                rwRow = setResponseAndWeight(_w, _resp, _weights, _respVec, rwRow, i);
+            }
+            rowHeaderPointer.set(_matrix._rowHeaders, nonZeroCount);
+        }
+    }
+
+    private static int initializeFromChunks(Chunk[] chunks, int weight, DataInfo di, long[][] rowHeaders, float[][] data, int[][] colIndex, int respIdx, float[] resp, float[] weights) {
+        int actualRows = 0;
+        int nonZeroCount = 0;
+        int rwRow = 0;
+
+        NestedArrayPointer rowHeaderPointer = new NestedArrayPointer();
+        NestedArrayPointer dataPointer = new NestedArrayPointer();
+
+        for (int i = 0; i < chunks[0].len(); i++) {
+            if (weight != -1 && chunks[weight].atd(i) == 0) continue;
+            actualRows++;
+            rowHeaderPointer.setAndIncrement(rowHeaders, nonZeroCount);
```

Comment:
How do we deal with all-zero rows ? The previous code had a nasty way to deal with it:

```java
                if(!foundNonZero){
                    foundNonZero = true;
                    for (int k = lastNonZeroRow; k < actualRows; k++) {
                        rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount;
                    }
                    lastNonZeroRow = actualRows;
                }
```

I don't see how setAndIncrement does this. I'd say this might be the reason for test failure: http://mr-0xc1:8080/blue/organizations/jenkins/h2o-3-pipeline/detail/PR-3541/3/tests

```
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: java.lang.ArrayIndexOutOfBoundsException: 119010
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at hex.tree.xgboost.matrix.SparseMatrixFactory$NestedArrayPointer.set(SparseMatrixFactory.java:88)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at hex.tree.xgboost.matrix.SparseMatrixFactory$InitializeCSRMatrixFromChunkIdsMrFun.map(SparseMatrixFactory.java:160)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at water.LocalMR.compute2(LocalMR.java:84)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at water.H2O$H2OCountedCompleter.compute(H2O.java:1420)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:974)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1477)
05-29 13:50:38.988 127.0.0.1:40000 203 FJ-1-121 ERRR: at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
```

Reply:
setAndIncrement removes the necessity for this if-clause, if you look closer you realize that

rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount

uses no new information gathered from the rows so we can set it immediatelly at the beginning of processing of a row

Reply:
I see. You're reight. The previous code could've been simplified to `            rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount;` at the beginning of the method.

Reply:
Then the pointer overflow must be originating somewhere else. :sob:

Reply:
bugfix incoming! 💇‍♂